src/java.logging/share/classes/java/util/logging/StreamHandler.java

```diff
@@ -199,19 +213,31 @@ public synchronized void publish(LogRecord record) {
         }
 
         try {
-            Writer writer = this.writer;
-            if (!doneHeader) {
-                writer.write(getFormatter().getHead(this));
-                doneHeader = true;
+            synchronized(this) {
+                Writer writer = this.writer;
+                if (!doneHeader) {
+                    writer.write(formatter.getHead(this));
+                    doneHeader = true;
+                }
+                writer.write(msg);
+                synchronousPostWriteHook();
             }
-            writer.write(msg);
         } catch (Exception ex) {
             // We don't want to throw an exception here, but we
             // report the exception to any registered ErrorManager.
             reportError(null, ex, ErrorManager.WRITE_FAILURE);
         }
     }
 
+    /**
+     * Overridden by other handlers in this package to facilitate synchronous
+     * post-write behaviour. If other handlers need similar functionality, it
+     * might be feasible to make this method protected (see JDK-8349206), but
+     * please find a better name if you do ;).
+     */
+    void synchronousPostWriteHook() {
+        // Empty by default.
+    }
```

Comment:
It would be good to have a comment making it clear that this method is called while holding the monitor on this. I know we have a test for that - but it could be good to make it clear in the code as well.
What about:
```suggestion
    void synchronousPostWriteHook() {
        // Empty by default. We could do:
        //    assert Thread.holdsLock(this);
        // but this is already covered by unit tests.
    }
```

Reply:
Done.